### PR TITLE
feat: Implement awala message endpoint and parse incoming parcel

### PR DESCRIPTION
**Pull request by levansuper**
closes: #9

### CODE REVIEW

File: src/server/routes/parcel.routes.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    'application/vnd.awala.parcel',
+    { parseAs: 'buffer' },
+    (_request, payload, next) => {
+      next(null, payload);
+    },
+  );
+
+  interface Request {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    readonly Body: Buffer;
```
Comment (levansuper) [inline src/server/routes/parcel.routes.ts:24]: this is needed for the typescript to read the body as a buffer

Comment (levansuper) [reply src/server/routes/parcel.routes.ts:24]: Shall I try and find a better way of doing this or go to the next task?

**Comment by levansuper** [comment]
Shall we move this part to a business logic file?
https://github.com/relaycorp/awala-endpoint-internet/pull/27/files#diff-7b82a13d28fd406ac3c2746ba9718b146995a508364131df83d203abe937e5cbR33-R57

File: src/server/routes/parcel.routes.spec.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      certificatePath.privateEndpoint,
+      keyPairSet,
+    );
+
+    const response = await server.inject({
+      ...validRequestOptions,
+      payload: parcelSerialized,
+    });
+
+    expect(response).toHaveProperty('statusCode', HTTP_STATUS_CODES.ACCEPTED);
```
Comment (gnarea) [inline src/server/routes/parcel.routes.spec.ts:56]: Log should be checked here too.

File: src/server/routes/parcel.routes.spec.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    const response = await server.inject({
+      ...validRequestOptions,
+
+      headers: {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        'Content-Type': 'application/json',
+      },
+    });
+
+    expect(response).toHaveProperty('statusCode', HTTP_STATUS_CODES.UNSUPPORTED_MEDIA_TYPE);
```
Comment (gnarea) [inline src/server/routes/parcel.routes.spec.ts:69]: Missing log check.

File: src/server/routes/parcel.routes.spec.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    const response = await server.inject({
+      ...validRequestOptions,
+      payload: parcelSerialized,
+    });
+
+    expect(response).toHaveProperty('statusCode', HTTP_STATUS_CODES.FORBIDDEN);
+    expect(JSON.parse(response.payload)).toHaveProperty(
+      'message',
+      'Parcel is well-formed but invalid',
+    );
```
Comment (gnarea) [inline src/server/routes/parcel.routes.spec.ts:110]: Missing log check.

File: src/testUtils/awala/parcel.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  readonly parcelSerialized: Buffer;
+  readonly parcel: Parcel;
+}
+async function generateStubNodeCertificate(
+  subjectPublicKey: CryptoKey,
+  issuerPrivateKey: CryptoKey,
+  options: Partial<{ readonly issuerCertificate: Certificate }> = {},
+): Promise<Certificate> {
+  const tomorrow = new Date();
+  tomorrow.setDate(tomorrow.getDate() + 1);
```
Comment (gnarea) [inline src/testUtils/awala/parcel.ts:22]: Let's please use `addDays()`.

File: package.json [hunk 6]
```diff
@@ -29,7 +30,8 @@
     "fastify": "^4.14.0",
     "fastify-plugin": "^4.5.0",
     "mongoose": "^7.0.5",
-    "pino": "^8.11.0"
+    "pino": "^8.11.0",
+    "uuid4": "^2.0.3"
```
Comment (gnarea) [inline package.json:34]: Where is `uuid4` used? 

If we do, in fact, use it, let's replace it with `randomUuid4()` from `node:crypto`.

File: package.json [hunk 7]
```diff
@@ -40,6 +42,7 @@
     "@types/jest": "^29.5.1",
     "@types/node": "^18.16.3",
     "@types/split2": "^4.2.0",
+    "@types/uuid4": "^2.0.0",
```
Comment (gnarea) [inline package.json:45]: Make sure to delete this too.

**Review by gnarea — CHANGES_REQUESTED** [review]

File: src/server/routes/parcel.routes.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    async handler(request, reply): Promise<void> {
+      let parcel;
+      try {
+        parcel = await Parcel.deserialize(bufferToArrayBuffer(request.body));
+      } catch (err) {
+        // Don't log the full error because 99.99% of the time the reason will suffice.
+        request.log.info({ reason: (err as Error).message }, 'Refusing malformed parcel');
+        return reply
+          .code(HTTP_STATUS_CODES.FORBIDDEN)
+          .send({ reason: 'Payload is not a valid RAMF-serialized parcel' });
```
Comment (gnarea) [inline src/server/routes/parcel.routes.ts:40]:
```suggestion
          .send({ message: 'Payload is not a valid RAMF-serialized parcel' });
```

File: src/server/routes/parcel.routes.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    async handler(request, reply): Promise<void> {
+      let parcel;
+      try {
+        parcel = await Parcel.deserialize(bufferToArrayBuffer(request.body));
+      } catch (err) {
+        // Don't log the full error because 99.99% of the time the reason will suffice.
+        request.log.info({ reason: (err as Error).message }, 'Refusing malformed parcel');
+        return reply
+          .code(HTTP_STATUS_CODES.FORBIDDEN)
```
Comment (gnarea) [inline src/server/routes/parcel.routes.ts:39]: I just noticed a bug in the original implementation in Pong: A malformed parcel should actually return a `400` response. Please fix it here.

File: src/server/routes/parcel.routes.ts (hunk 1 above)
Comment (gnarea) [reply src/server/routes/parcel.routes.ts:24]: The only problem I see with this is that you're redeclaring an interface every time that function is called. I'd just move it outside the function and call it something like `BinaryRequest`.

Comment (levansuper) [reply src/server/routes/parcel.routes.ts:24]: In my comment, I was referring to the disabling the lint. If that's ok I'll just implement it exactly the same way as in the example

Comment (gnarea) [reply src/server/routes/parcel.routes.ts:24]: OK, that was ambiguous then, but I understand it now. Yes, please do what I did in Pong.

File: src/server/routes/parcel.routes.spec.ts (hunk 3 above)
Comment (levansuper) [reply src/server/routes/parcel.routes.spec.ts:69]: It's checked by the contentTypeParser and is automatically declined by fastify. Shall we try and inject logger?

Comment (gnarea) [reply src/server/routes/parcel.routes.spec.ts:69]: I see. In that case, let's ignore my comment.

**Review by gnarea — APPROVED** [review]
Thanks!

**Pull request merged**